component/azstorage/file_share.go

```diff
@@ -508,28 +573,162 @@ func (fs *FileShare) List(prefix string, marker *string, count int32) ([]*intern
 	return fileList, listFile.NextMarker.Val, nil
 }
 
+// ReadToFile : Download an Azure file to a local file
 func (fs *FileShare) ReadToFile(name string, offset int64, count int64, fi *os.File) error {
-	return syscall.ENOTSUP
+	log.Trace("FileShare::ReadToFile : name %s, offset : %d, count %d", name, offset, count)
+	//defer exectime.StatTimeCurrentBlock("FileShare::ReadToFile")()
+
+	if offset != 0 {
+		log.Err("FileShare::ReadToFile : offset is not 0")
+		return errors.New("offset is not 0")
+	}
+
+	fileName, dirPath := getFileAndDirFromPath(filepath.Join(fs.Config.prefixPath, name))
+	fileURL := fs.Share.NewDirectoryURL(dirPath).NewFileURL(fileName)
+
+	defer log.TimeTrack(time.Now(), "FileShare::ReadToFile", name)
+	_, err := azfile.DownloadAzureFileToFile(context.Background(), fileURL, fi, fs.downloadOptions)
+
+	if err != nil {
+		e := storeFileErrToErr(err)
+		if e == ErrFileNotFound {
+			return syscall.ENOENT
+		} else {
+			log.Err("FileShare::ReadToFile : Failed to download file %s (%s)", name, err.Error())
+			return err
+		}
+	}
+
+	return nil
 }
 
+// ReadBuffer : Downloads a file to a buffer
 func (fs *FileShare) ReadBuffer(name string, offset int64, len int64) ([]byte, error) {
-	return nil, syscall.ENOTSUP
+	log.Trace("FileShare::ReadBuffer : name %s", name)
+	var buff []byte
+
+	if offset != 0 {
+		log.Err("FileShare::ReadToFile : offset is not 0")
+		return buff, errors.New("offset is not 0")
+	}
+
+	if len == 0 {
+		len = azfile.CountToEnd
+		attr, err := fs.GetAttr(name)
+		if err != nil {
+			return buff, err
+		}
+		buff = make([]byte, attr.Size)
+	} else {
+		buff = make([]byte, len)
+	}
+
+	fileName, dirPath := getFileAndDirFromPath(filepath.Join(fs.Config.prefixPath, name))
+	fileURL := fs.Share.NewDirectoryURL(dirPath).NewFileURL(fileName)
+
+	_, err := azfile.DownloadAzureFileToBuffer(context.Background(), fileURL, buff, fs.downloadOptions)
+
+	if err != nil {
+		e := storeFileErrToErr(err)
+		if e == ErrFileNotFound {
+			return buff, syscall.ENOENT
+		} else if e == InvalidRange {
+			return buff, syscall.ERANGE
+		}
+
+		log.Err("FileShare::ReadBuffer : Failed to download file %s (%s)", name, err.Error())
+		return buff, err
+	}
+
+	return buff, nil
 }
 
+// ReadInBuffer : Download specific range from a file to a user provided buffer
 func (fs *FileShare) ReadInBuffer(name string, offset int64, len int64, data []byte) error {
-	return syscall.ENOTSUP
+	log.Trace("FileShare::ReadInBuffer : name %s", name)
+
+	if offset != 0 {
+		log.Err("FileShare::ReadToFile : offset is not 0")
```

Comment:
This is the `ReadInBuffer` method not `ReadToFile`. Also, ReadInBuffer supports offsets that are not 0. what's the reason for this check?

Reply:
Oops I'll fix that. 
ReadInBuffer() is supposed to support offsets that are not 0 but the SDK doesn't. ReadInBuffer() calls DownloadAzureFileToBuffer() in the SDK, which doesn't take offset as one of the arguments (or as part of any of the arguments).